cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrust/transform.h>
+#include <rmm/thrust_rmm_allocator.h>
+
+#include <cudf/cudf.h>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {    // anonymous
+
+/* --------------------------------------------------------------------------*/
+/**
+* @brief Functor called by the `type_dispatcher` in order to perform a copy if/else
+*        using a filter function to select from lhs/rhs columns.
+*/
+/* ----------------------------------------------------------------------------*/
+struct copy_if_else_functor {
+   template <typename T, typename Filter>
+   void operator()(  Filter filter,
+                     column_view const& lhs,
+                     column_view const& rhs,
+                     mutable_column_view& out,
+                     cudaStream_t stream)
+   {
+      auto begin  = thrust::make_zip_iterator(thrust::make_tuple( thrust::make_counting_iterator(0),
+                                                                  lhs.begin<T>(),
+                                                                  rhs.begin<T>()));
+
+      auto end  = thrust::make_zip_iterator(thrust::make_tuple(thrust::make_counting_iterator(lhs.size()),
+                                                               lhs.end<T>(),
+                                                               rhs.end<T>()));
+      
+      thrust::transform(rmm::exec_policy(stream)->on(stream), begin, end, out.begin<T>(),
+                        [filter] __device__ (thrust::tuple<size_type, T, T> i)
+                        {
+                           return filter(thrust::get<0>(i)) ? thrust::get<1>(i) : thrust::get<2>(i);
+                        });
+   } 
+};
+
+}  // anonymous namespace
+
+/**
+ * @brief   Returns a new column, where each element is selected from either @p lhs or 
+ *          @p rhs based on the filter lambda. 
+ * 
+ * @p filter must be a functor or lambda with the following signature:
+ * bool __device__ operator()(cudf::size_type i);
```

Comment:
Instead of a functor, this should be an _iterator_ that resolves to `true/false` for each element.